cql3/statements/select_statement.cc

```diff
@@ -404,26 +405,52 @@ select_statement::do_execute(query_processor& qp,
 
     auto key_ranges = _restrictions->get_partition_key_ranges(options);
 
+    bool send_tablet = false;
+
+    auto token = dht::token();
+    unsigned shard;
+
+    if (key_ranges.size() == 1 && query::is_single_partition(key_ranges.front())) {
+        token = key_ranges[0].start()->value().as_decorated_key().token();
+        shard = _schema->table().shard_of(token);
+    }
+
+    auto&& table = qp.proxy().local_db().find_column_family(_schema);
+    auto erm = table.get_effective_replication_map();
+    locator::tablet_replica_set tablet_replicas = {};
+    auto token_range = std::pair<dht::token, dht::token>();
+
+    if (!locator::contains(tablet_replicas, erm->get_token_metadata().get_my_id()) || this_shard_id() != shard) {
+        tablet_replicas = erm->get_replicas_for_reading(token);
+        token_range = erm->get_token_range(token);
+    }
+
     if (db::is_serial_consistency(options.get_consistency())) {
         if (key_ranges.size() != 1 || !query::is_single_partition(key_ranges.front())) {
              throw exceptions::invalid_request_exception(
                      "SERIAL/LOCAL_SERIAL consistency may only be requested for one partition at a time");
         }
-        unsigned shard = _schema->table().shard_of(key_ranges[0].start()->value().as_decorated_key().token());
         if (this_shard_id() != shard) {
             return make_ready_future<shared_ptr<cql_transport::messages::result_message>>(
                     qp.bounce_to_shard(shard, std::move(const_cast<cql3::query_options&>(options).take_cached_pk_function_calls()))
                 );
         }
     }
 
+    auto f = make_ready_future<shared_ptr<cql_transport::messages::result_message>>();
+
     if (!aggregate && !_restrictions_need_filtering && (page_size <= 0
             || !service::pager::query_pagers::may_need_paging(*_schema, page_size,
                     *command, key_ranges))) {
-        return execute_without_checking_exception_message_non_aggregate_unpaged(qp, command, std::move(key_ranges), state, options, now);
+        f = execute_without_checking_exception_message_non_aggregate_unpaged(qp, command, std::move(key_ranges), state, options, now);
     } else {
-        return execute_without_checking_exception_message_aggregate_or_paged(qp, command, std::move(key_ranges), state, options, now, page_size, aggregate, nonpaged_filtering);
+        f = execute_without_checking_exception_message_aggregate_or_paged(qp, command, std::move(key_ranges), state, options, now, page_size, aggregate, nonpaged_filtering);
     }
+
+    return f.then([tablet_replicas = std::move(tablet_replicas), token_range] (auto res) mutable {
+        res->add_tablet_info(std::move(tablet_replicas), token_range);
+        return res;
+    });
 }
```

Comment:
This should be conditional. If we're on the right node/shard, no need to spam the driver and no need to add a continuation.

It should also be conditional on it being a prepared statement with partition key information. Otherwise there is no way for the driver to send the request to the right node/shard, and we're just spamming it.

Reply:
If we are on right node/shard the `tablet_replicas` is default and `add_tablet_info()` does nothing, should I still add a condition here?

Reply:
Not call future::then. It's quite expensive as it adds an allocation and a task. You should be able to see it in perf-simple-query.

Reply:
Done